up-streamer/src/ustreamer.rs

```diff
@@ -0,0 +1,986 @@
+/// assert!(streamer
+///     .add_forwarding_rule(local_route.clone(), local_route.clone())
+///     .await
+///     .is_err());
+///
+/// // Rule already exists so it should report an error
+/// assert!(streamer
+///     .add_forwarding_rule(local_route.clone(), remote_route.clone())
+///     .await
+///     .is_err());
+///
+/// // Try and remove an invalid rule
+/// assert!(streamer
+///     .delete_forwarding_rule(remote_route.clone(), remote_route.clone())
+///     .await
+///     .is_err());
+///
+/// // remove valid routing rules
+/// assert_eq!(
+///     streamer
+///         .delete_forwarding_rule(local_route.clone(), remote_route.clone())
+///         .await,
+///     Ok(())
+/// );
+/// assert_eq!(
+///     streamer
+///         .delete_forwarding_rule(remote_route.clone(), local_route.clone())
+///         .await,
+///     Ok(())
+/// );
+///
+/// // Try and remove a rule that doesn't exist, should report an error
+/// assert!(streamer
+///     .delete_forwarding_rule(local_route.clone(), remote_route.clone())
+///     .await
+///     .is_err());
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct UStreamer {
+    #[allow(dead_code)]
+    name: String,
+    registered_forwarding_rules: ForwardingRules,
+    transport_forwarders: TransportForwarders,
+    transport_forwarder_count: TransportForwarderCount,
+    in_transport_out_authorities: InTransportOutAuthorities,
+    message_queue_size: usize,
+}
+
+impl UStreamer {
+    /// Creates a new UStreamer which can be used to add forwarding rules.
+    ///
+    /// # Parameters
+    ///
+    /// * name - Used to uniquely identify this UStreamer in logs
+    /// * message_queue_size - Determines size of channel used to communicate between `ForwardingListener`
+    ///                        and the worker tasks for each currently routed `UTransport`
+    pub fn new(name: &str, message_queue_size: usize) -> Self {
+        let name = format!("{USTREAMER_TAG}:{name}:");
+        // Try to initiate logging.
+        // Required in case of dynamic lib, otherwise no logs.
+        // But cannot be done twice in case of static link.
+        let _ = env_logger::try_init();
+        debug!(
+            "{}:{}:{} UStreamer created",
+            &name, USTREAMER_TAG, USTREAMER_FN_NEW_TAG
+        );
+
+        Self {
+            name: name.to_string(),
+            registered_forwarding_rules: Arc::new(Mutex::new(HashMap::new())),
+            transport_forwarders: Arc::new(Mutex::new(HashMap::new())),
+            transport_forwarder_count: Arc::new(Mutex::new(HashMap::new())),
+            in_transport_out_authorities: Arc::new(Mutex::new(HashMap::new())),
+            message_queue_size,
+        }
+    }
+
+    fn uauthority_to_uuri(authority: UAuthority) -> UUri {
+        UUri {
+            authority: Some(authority).into(),
+            ..Default::default()
+        }
+    }
+
+    #[inline(always)]
+    fn forwarding_id(r#in: &Route, out: &Route) -> String {
```

Comment:
Might be more readable to replace the raw identifier with something like input_route.

Reply:
I suppose so, but end-users of the library won't really need to use / reference it. I'll keep as-is for now. Thx for feedback.